### DISCUSSION TITLE
Add example to wiki of writing simple bookmark functions for modes not having them

### DISCUSSION THREAD

**Discussion by amno1** [alphapapa/activities.el#39] [Q&A]
Is there some before/after hook I can hook a function into before the bookmark is opened? 

I would like  to restore a Sly session, with sbcl and a custom saved image. I am thinking of:

    open bookmarked asdf file or at least project directory
    start sly with a given compiler (I have different builds of sbcl) and saved core image

I haven't looked much at the code; I have just installed it and played a bit with it, so I am not very familiar with it. I see the activities--bookmark-buffer function.;  or in activites--serialize/deserialize?

I could then add something to before (deserialize) hook to run Sly and wait until repl is up and running.

Or perhaps there is already a better place or a better way?

**Comment by alphapapa** [comment]
The way to handle a buffer whose major mode doesn't support bookmarks is to write bookmark support for it.  It's actually very simple.  Here's an example of a user doing that for another mode:

https://github.com/alphapapa/activities.el/issues/19#issuecomment-1925700528

You could also make a bookmark "from scratch" and set the handler function to one that does whatever setup you need; what matters is that the buffer to be displayed is current when the function returns.

**Comment by amno1** [comment]
I haven't tested much; but as I understand, bookmarks support only file-backed buffers We can't bookmark *scratch* buffer for example . 

We can perhaps let-bind buffer-file-name  in a non-file buffer to something in /var/tmp or some other trick to fake it, but as I see it, it is a fundamental limitation of using bookmarks as the framework on which to implement this.

Recreating the state of the system is more than just restoring buffers and window configuration.  But restoring buffers and window configuration is certainly a part of it, and useful on its own. I don't think your library itself can solve the entire problem without support for better state dumping. That perhaps needs to be implemented in the core. I don't know, to be honest. This library have a place, but it would be helpful if we get a hook that runs before and after so we can setup the system, for example start some external processes, load files into them etc. 

For example, Sly repl buffer does not exist before Sly is started and connected to a running Lisp process. It means I need to hook code to start sbcl process and connect to it if associated asdf/project/file does not already have a connection, before switching to it. It would be also useful if I can hook a function that asks sly to instruct the running sbcl to save a new image when exiting (save-lisp-and-die), so I don't have to do it manually.

What you suggest with making bookmark from scratch is not unreasonable, but you are using bookmark to open activity. How is it suppose to work; a two layer bookmark sort of? It feels we are getting into waters of Bookmark+, which is rather an anti-pattern in my eyes. Isn't it better to provide a hook before bookmark is called, and after it has saved to setup and tear down things?

I had quite bad flu, and those few moments I was at computer, went to something else, so I haven't looked at this much at all, and basically nothing at your code, so I have no suggestions yet.  This is just my understanding thus far.

**Comment by amno1** [comment]
After using activities for a while, I think I understand a problem a bit better. This is not about writing bookmark handlers. This is different than bookmark support, since bookmarks are per mode. I wish to be able to have different activities involving the same modes but different processes. Bookmarks are not really convenient way to work with this.

In my concrete example, I setup terminal with SBCL process, as one of setup steps, and connect from Emacs via Sly to that project. I do that for different projects. For example, one problem here is that Sly will create a repl with certain name. I wish to rename a repl with some project specific name so I can switch easier to it when I have multiple repls open. Now I will have to invent some kind of framework so my bookmark handler for sly-mrepl-mode can figure out a name of the project and do what I want, and suddenly we are in a waters of Bookmark+. Also, what if I don't want to rename my repl, and would like to do something else? Where to even start, write a handler for Dired so it runs correct thing when I switch to certain folder? Use dir-locals?  I don't know honestly, nothing seems to be convenient and simple enough. I have written a simple command just startup project and setup things, and I use activities to resume if I switch away and work with something else.

In some different activity, that might involve the very same tools and programs, I might wish to do something else, run another setup process, or perhaps I wish to run another Lisp implementation for example. I don't think that is not something bookmarks alone can solve, at least not conveniently. The problem is of course in starting various processes and setting up things, so they *can* be bookmarked. 

Of course, my goal here is not to rant. The problem is solvable; as said I currently have my own script that setups stuff before I can use activities, but it would be much more pragmatic with some kind of a hook for per-activity setup so I can hook my command into it, or something similar.  Think of projectile-switch-to-project function, but activity specific so to say. Admittedly, this is looking at Activities a little bit like a project manager, but I think it is a natural regression. 

The only problem I am not sure how to do nicely with activities.el is how to associate a hook with an activity in a pretty way. I am not sure asking interactively, in activity-new is the prettiest way to do it, or one should perhaps do as some kind of specially named function or what would be a nice workflow for that.

Thoughts? Is there some simpler way to automate this I don't see?

**Reply by alphapapa** [reply]
Having some kind of on-activation hook for activities is not a bad idea.  It would probably not have much in the way of UI, but Lisp users could set it up in their config.
